src/mongocxx/instance.cpp

```diff
@@ -14,28 +14,63 @@
 
 #include <mongocxx/instance.hpp>
 
+#include <mongoc.h>
+#include <utility>
+
 #include <bsoncxx/stdx/make_unique.hpp>
 
-#include <mongoc.h>
+#include <mongocxx/private/libmongoc.hpp>
+#include <mongocxx/logger.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
 
 namespace {
-void log_handler(mongoc_log_level_t, const char *, const char *, void *) {
+extern "C" void null_log_handler(::mongoc_log_level_t, const char *, const char *, void *) {
 }
+
+log_level convert_log_level(::mongoc_log_level_t mongoc_log_level) {
+    switch (mongoc_log_level) {
+        case MONGOC_LOG_LEVEL_ERROR:
+            return log_level::k_error;
+        case MONGOC_LOG_LEVEL_CRITICAL:
+            return log_level::k_critical;
+        case MONGOC_LOG_LEVEL_WARNING:
+            return log_level::k_warning;
+        case MONGOC_LOG_LEVEL_MESSAGE:
+            return log_level::k_message;
+        case MONGOC_LOG_LEVEL_INFO:
+            return log_level::k_info;
+        case MONGOC_LOG_LEVEL_DEBUG:
+            return log_level::k_debug;
+        case MONGOC_LOG_LEVEL_TRACE:
+            return log_level::k_trace;
+        default:
+            // TODO: MONGOCXX_UNREACHABLE (CXX-628)
+            std::abort();
+    }
 }
 
-namespace mongocxx {
-MONGOCXX_INLINE_NAMESPACE_BEGIN
+extern "C" void user_log_handler(::mongoc_log_level_t mongoc_log_level, const char* log_domain,
+                                 const char* message, void* user_data) {
+    (*static_cast<logger*>(user_data))(convert_log_level(mongoc_log_level),
```

Comment:
I think that you need to either:
- wrap this in try catch and do something on an exception, so that you don't throw through C context, OR
- require that the handler operator() be noexcept